### PR TITLE
feat : dependabot workflow automation for updating dependency

### PR DESCRIPTION
**Pull request by Rajpratik71**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Dependabot workflow automation for updating dependency

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
feat : dependabot workflow automation for updating dependency
```

Signed-off-by: Pratik Raj <rajpratik71@gmail.com>

### CODE REVIEW

File: .github/dependabot.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+version: 2
+updates:
+  # Dependencies listed in go.mod
+  - package-ecosystem: "gomod"
+    directory: "/" # Location of package manifests
+    schedule:
+      interval: "weekly"
```
Comment (mboersma) [inline .github/dependabot.yml:7]: Maybe we should add some labels here too?

```yaml
labels:
  - "kind/cleanup"
  - "priority/important-soon"
  - "wg/security-audit"
```

Comment (Rajpratik71) [reply .github/dependabot.yml:7]: Make sense.

Beside these labels any other label to add ?

Comment (mboersma) [reply .github/dependabot.yml:7]: Those were the only ones I saw that seemed like good defaults.

Comment (Rajpratik71) [reply .github/dependabot.yml:7]: ok. 

For "go.mod" adding all these, but for "github-action" only adding "kind/cleanup"

**Review by mboersma — COMMENTED** [review]
/lgtm
/label tide/merge-method-squash

I think this is a valuable addition--a weekly PR that updates Go and GH Actions dependencies--but let's see what other maintainers think.

File: .github/dependabot.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    labels:
```
Comment (CecileRobertMichon) [inline .github/dependabot.yml:8]: we might want to add the ok-to-test label here like CAPI does so we don't have to approve each PR manually for testing https://github.com/kubernetes-sigs/cluster-api/blob/main/.github/dependabot.yml#L10

File: .github/dependabot.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+version: 2
```
Comment (CecileRobertMichon) [inline .github/dependabot.yml:1]: CAPI also has an associated GitHub action that updates generated code, I believe we need something equivalent here https://github.com/kubernetes-sigs/cluster-api/blob/main/.github/workflows/dependabot.yml

**Comment by nawazkh** [comment]
Thank you for adding this! 
Quicklinks if someone is looking for `Depandabot` functionalities:
- [dependabot-with-github-actions](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/automating-dependabot-with-github-actions)
- [dependabot-update-script](https://github.com/dependabot/dependabot-script#dependabot-update-script)

**Comment by CecileRobertMichon** [comment]
/ok-to-test

How do we validate that this works prior to merging?

**Comment by Rajpratik71** [comment]
> /ok-to-test
> 
> How do we validate that this works prior to merging?

Merging has no direct impact or will not modify anything in the code.

As it is a GitHub workflow, scheduled to run "weekly", 

weekly it will raise PRs for dependency updates, which will be tested by CI.

Once, the CI build and tests are successful those PRs can be merged.

This will reduce maintenance efforts for dependency.

File: workflows/dependabot.yml [hunk 4]
```diff
@@ -0,0 +1,46 @@
+name: dependabot
```
Comment (nawazkh) [inline workflows/dependabot.yml:1]: Was this file supposed to be created in `$PROJECT_ROOT/.github/workflows/` dir than in `$PROJECT_ROOT/workflows/` dir?

File: .github/dependabot.yml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+updates:
+  # Dependencies listed in go.mod
+  - package-ecosystem: "gomod"
+    directory: "/" # Location of package manifests
+    schedule:
+      interval: "weekly"
+    commit-message:
+        prefix: ":seedling:"
```
Comment (nawazkh) [inline .github/dependabot.yml:9]: Would it be a good idea to include `include: "scope"` in `commit-message` to get a list of dependencies updated in the commit? Reference: [commit-message](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#commit-message)

File: .github/dependabot.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+version: 2
+updates:
+  # Dependencies listed in go.mod
+  - package-ecosystem: "gomod"
+    directory: "/" # Location of package manifests
+    schedule:
+      interval: "weekly"
```
Comment (nawazkh) [inline .github/dependabot.yml:7]: Do we fancy a day to check for new versions? Say Wednesday? 
The default behavior will make depandabot _check at random time on Monday_:
Referring: [schedule.day](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#scheduleday)

File: workflows/dependabot.yml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+          ${{ runner.os }}-go-
+    - name: Update all modules
+      run: make generate-modules
+    - name: Update generated code
+      run: make generate
+    - uses: EndBug/add-and-commit@61a88be553afe4206585b31aa72387c64295d08b # tag=v9.1.1
+      name: Commit changes
+      with:
+        author_name: dependabot[bot]
+        author_email: 49699333+dependabot[bot]@users.noreply.github.com
```
Comment (nawazkh) [inline workflows/dependabot.yml:44]: I am curious to know if `author_email` was auto-generated? If yes, how?

**Review by nawazkh — COMMENTED** [review]
Just some thoughts and doubts.

**Comment by CecileRobertMichon** [comment]
Some changes just merged in CAPI on the configuration, could you please apply the same changes?

https://github.com/kubernetes-sigs/cluster-api/pull/7632/files

File: workflows/dependabot.yml (hunk 4 above)
Comment (CecileRobertMichon) [reply workflows/dependabot.yml:1]: +1

File: .github/dependabot.yml (hunk 6 above)
Comment (CecileRobertMichon) [reply .github/dependabot.yml:7]: I think random time on Monday is fine to start, we can tweak it later if needed

File: .github/dependabot.yml (hunk 5 above)
Comment (CecileRobertMichon) [reply .github/dependabot.yml:9]: +1 

Also CAPZ doesn't use the "seedling"  prefix like CAPI does, we just the k8s release note generator tool which uses the labels to classify release notes

**Comment by CecileRobertMichon** [comment]
@Rajpratik71 are you still interested in getting this over the finish line for the v1.7 milestone??

**Comment by Rajpratik71** [comment]
@CecileRobertMichon  made the changes a/c to comments

File: workflows/dependabot.yml (hunk 7 above)
Comment (nawazkh) [reply workflows/dependabot.yml]: I could not find any correlation to `49699333+dependabot[bot]@users.noreply.github.com` to dependabot. 

Dependabot's github page publicizes its email id as [support@github.com](mailto:support@github.com)
Screenshot 2022-12-09 at 2 21 14 PM

one of the upgrade commit in my fork of CAPZ has dependabot's email ID as `support@github.com` too. Refer https://github.com/nawazkh/cluster-api-provider-azure/pull/1/commits/b7f7af4bf7cefa40c1527f2f281482cd955fe4a5

As this is just meta data for us repo watchers, I am ok with either `author_email: 49699333+dependabot[bot]@users.noreply.github.com` or `support@github.com`.

Comment (nawazkh) [reply workflows/dependabot.yml]: Feel free to collapse this comment thread as resolved!

**Comment by nawazkh** [comment]
/lgtm
Thanks! 🚢

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

let's merge this before the weekend and see what happens on Monday :) thanks @Rajpratik71!

**Pull request merged**